### PR TITLE
gamess-us: init at 2021R2P1

### PR DESCRIPTION
**Pull request by sheepforce**
This is an attempt to add the current GAMESS-US distribution. I've had it working with `nixos-20.09` in my own overlay some time ago. Now building fails during compilation. A script seems simply to fail but I can't figure out why. It fails at `make modules` with `sed: can't read modules_dft: No such file or directory`. I've noticed that you GAMESS-US in the overlay some time ago. @markuskowa could you maybe help me here?

### CODE REVIEW

**Comment by markuskowa** [comment]
I can take a look next week. 
I removed Gamess since it failed to build at some point and we were not using it (and the custom build system was hard to deal with).

**Comment by sheepforce** [comment]
> I removed Gamess since it failed to build at some point

GAMESS is super sensitive against the compiler versions. They have separate compiler optimisation flags for each minor version of gfortran. And it usually fails if the compiler is newer than what they have optimisation flags for. I guess we could use `gcc9Stdenv` and a fixed gfortran version (`gfortran9` if it would be available)?

File: gamess-us/default.nix [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+      # Copy auxdata, which contains parameters and basis sets
+      cp -r auxdata $out/share/gamess/.
+
+      # Copy the test files
+      cp -r tests $out/share/gamess/.
+
+      runHook postInstall
+    '';
+
+    binSearchPath = lib.strings.makeSearchPath "bin" [ openssh mpi tcsh hostname ];
```
Comment (markuskowa) [inline gamess-us/default.nix:195]: That should be a `let` in `postFixup`

**Comment by markuskowa** [comment]
> and a fixed gfortran version (gfortran9 if it would be available)?

There are plenty of choices from 4.8 - 10 (see https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&query=gfortran).

File: gamess-us/default.nix [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+  (!blas.isILP64)
+  "A 32 bit integer implementation of BLAS is required.";
+
+let
+  blasImplementation = blas.passthru.implementation;
+  blasProvider = blas.passthru.provider;
+
+  version = "2020R2";
+  gfortranVersion = lib.versions.majorMinor gfortran.version;
+  libxcSource = fetchFromGitLab {
```
[the remark this answers is not in the document]
Comment (sheepforce) [reply gamess-us/default.nix:20]: I do not fully understand what's going on in the GAMESS-US build system. Especially I haven't figured out where it links to LibXC. But using `libxc.src` is a good idea.

File: gamess-us/default.nix [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+assert
+  lib.asserts.assertMsg
+  (!blas.isILP64)
+  "A 32 bit integer implementation of BLAS is required.";
+
+let
+  blasImplementation = blas.passthru.implementation;
+  blasProvider = blas.passthru.provider;
+
+  version = "2020R2";
```
[the remark this answers is not in the document]
Comment (sheepforce) [reply gamess-us/default.nix:18]: Yes, it is. The version from the beginning of 2020 was 2020R1, i think? Or are there multiple versions of 2020R2?

File: gamess-us/default.nix [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+      blas
+    ];
+
+    propagatedBuildInputs = [
+      tcsh
+      mpi
+    ];
+
+    src = requireFile rec {
+      name = "gamess-current.tar.gz";
```
[the remark this answers is not in the document]
Comment (sheepforce) [reply gamess-us/default.nix:102]: That's true. It always downloads as `gamess-current.tar.gz` and I haven't found an option to get access to the older versions unfortunately. But `nix-store --add-fixed sha256 ...` requires matching filenames, doesn't it? So after download one would need to rename it. But still, we should probably follow your suggestion.

File: gamess-us/default.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (markuskowa) [reply gamess-us/default.nix:18]: The latest version on their download page is `GAMESS version September 30, 2020 R2`, which also fits the sha256 you have here.

File: gamess-us/default.nix (hunk 4 above)
[the remark this answers is not in the document]
Comment (markuskowa) [reply gamess-us/default.nix:102]: Gamess may not offer old versions. However, we start now `09-2020-R2` and we put the file in a folder at `srcurl`. With the next update we have to put a new file in `srcurl`, which then unfortunately has the same name. If you rename the file to include the version, you will be able to re-build a previous version for which you most likely still have `gamess-MM-YYYY-RN.tar.gz` (let's use this naming schem) in `srcurl`.

**Comment by markuskowa** [comment]
Have you seen [this](https://github.com/gentoo/sci/tree/master/sci-chemistry/gamess)? Maybe that there are some good ideas in there?

File: gamess-us/default.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (sheepforce) [reply gamess-us/default.nix]: Dropped LibXC for now. It complicates things and I guess DFT is not GAMESS primary strength.

**Comment by sheepforce** [comment]
I have finally managed to build it, it works! :partying_face:

**Comment by markuskowa** [comment]
Very nice! Do have a short input at hand that we can use for a test?

**Comment by markuskowa** [comment]
Looks good. Should I merge it?

**Comment by sheepforce** [comment]
Yes, the test runs fine and everything should be working now. Altough I haven't tested other MPI implementations.

**Comment by markuskowa** [comment]
> Yes, the test runs fine and everything should be working now. Altough I haven't tested other MPI implementations.

I have "big" jobset on my internal Hydra, which builds different MPI and BLAS implementations.

**Pull request merged**